and-cli-deploy-azure-web-app.js

```diff
@@ -0,0 +1,184 @@
+#!/usr/bin/env node
+require("./command-runner").run(async () => {
+    // -----------------------------------------------------------------------------------------
+    // #region Imports
+    // -----------------------------------------------------------------------------------------
+
+    const azure         = require("./_modules/azure");
+    const dotnetPath    = require("./_modules/dotnet-path");
+    const echo          = require("./_modules/echo");
+    const program       = require("commander");
+    const shell         = require("shelljs");
+
+    // #endregion Imports
+
+    // -----------------------------------------------------------------------------------------
+    // #region Variables
+    // -----------------------------------------------------------------------------------------
+
+    let   appName             = null;
+    let   branch              = null;
+    let   clientId            = null;
+    let   force               = false;
+    const pythonInstallerUrl  = "https://www.python.org/ftp/python/3.7.4/python-3.7.4-amd64.exe";
+    let   remote              = null;
+    let   resourceGroup       = null;
+    let   secret              = null;
+    let   tenantId            = null;
+    let   username            = null;
+
+    // #endregion Variables
+
+    /// -----------------------------------------------------------------------------------------
+    // #region Functions
+    // -----------------------------------------------------------------------------------------
+
+    const deployAzureWebApp = {
+        createRemoteIfMissing() {
+            if (shell.exec(`git remote get-url ${remote}`).code !== 0) {
+                const url = shell.exec(`az webapp deployment list-publishing-credentials --name ${appName} --resource-group ${resourceGroup} --query scmUri --output tsv`);
+
+                if (shell.exec(`git remote add ${remote} ${url}`).code !== 0)
+                {
+                    echo.error("Error trying to add remote!");
+                    azure.logout();
+                    shell.exit(1);
+                }
+            }
+        },
+        pushToRemote() {
+            let pushCmd = `git push ${remote} ${branch}:master`;
+
+            if (force) {
+                pushCmd += " -f";
+            }
+
+            if (shell.exec(pushCmd).code !== 0) {
+                echo.error(" - Failed pushing to Web App remote");
+                azure.logout();
+                shell.exit(1);
+            }
+        },
+        description() {
+            return `Pushes indicated branch from the current git repo to indicated Azure Web App git repo, which then deploys to configured Azure Web App environment`;
+        },
+        async run() {
+            // Check system/command requirements
+            this.validateOrExit();
+
+            // Login to Azure
+            if (username != null) {
```

Comment:
Thought: With the modules of the CLI, make sure to consider with a given function whether or not you can see any use cases where consumers would like to get the error code back and decide if it should do a shell.exit. While we do have scenarios where the module functions will exit, there are a lot where we are just returning the result and allowing the outer commands to decide what to do.

Reply:
Yea, that certainly makes sense as something to consider. In this case, I feel like logging in/out of Azure is such a key command pre/post azure cli operations that it makes sense to just exit if those fail. I can't think of a case where the consumer would want do anything besides exit beyond *maybe* showing a more detailed error message.